lib/cloud_text.rb

```diff
@@ -9,47 +9,64 @@ def self.clean_text(input, options = {})
+    stemming_enabled = @options.fetch(:stemming, false)
 
-    # Lowercase all tokens
+    # Get variables
     @language = @options.fetch(:language, "en")
-    @input = UnicodeUtils.downcase(@input, @language.to_sym)
-
-    # Filter stopwords
-    @cutsom_stopwords = @options.fetch(:stopwords, [])
-    if @language == "tr"
-      stopword_filter = Stopwords::Filter.new tr_stopwords
-    else
-      stopword_filter = Stopwords::Snowball::Filter.new @language
-    end
+    @custom_stopwords = @options.fetch(:stopwords, [])
+
+    @input = process_text(@input, stemming_enabled, remove_digits, @language, @custom_stopwords)
+    count_words(@input)
+  end
+
+  private
+
+  def self.process_text(input, stemming_enabled, remove_digits, language, custom_stopwords)
+    input = remove_punctuation(input, remove_digits)
+    input = reduce_whitespaces(input)
+    input = lowercase_words(input, language)
+    input = filter_stopwords(input, language, custom_stopwords) # Get custom_stopwords from user and filter words
+    input = stemming(input, language) if stemming_enabled
+  end
+
+  def self.remove_punctuation(input, remove_digits)
+    regex = remove_digits ? /[^A-Za-z0-9^şŞıİçÇöÖüÜĞğ\s]|_|\d/ : /[^A-Za-z0-9^şŞıİçÇöÖüÜĞğ\s]|_/
+    input.gsub(regex, ' ')
+  end
+  
+  # Reduce multiple whitespaces into single whitespace
+  def self.reduce_whitespaces(input)
+    input.gsub(/\s+/, ' ')
+  end
+  
+  def self.lowercase_words(input, language)
+    UnicodeUtils.downcase(input, language.to_sym)
+  end
+
+  # Remove stopwords for given language and also given custom stopwords
+  def self.filter_stopwords(input, language, custom_stopwords = nil)
+    stopword_filter = language == "tr" ? Stopwords::Filter.new(tr_stopwords) : stopword_filter = Stopwords::Snowball::Filter.new(language)
 
     # Here we intentionally do not downcase custom_stopwords 
     # since we want to filter only capitalized version of a word
-    stopword_filter.stopwords << @cutsom_stopwords
-    @input = stopword_filter.filter(@input.split)
-    
-    # Stemming
-    stemming_enabled = @options.fetch(:stemming, false)
-    if stemming_enabled
-      stemmer = Lingua::Stemmer.new(language: @language)
-      @input = @input.map do |word|
-        stemmer.stem(word)
-      end
-    end
+    stopword_filter.stopwords << custom_stopwords if custom_stopwords
+    stopword_filter.filter(input.split)
+  end
 
-    # Counting the words, generate array for each element like => [word, frequency]
-    @result = @input.each_with_object(Hash.new(0)) { |token, hash| hash[token] += 1 }.sort_by {|k,v| v}.reverse
+  def self.stemming(input, language)
+    stemmer = Lingua::Stemmer.new(language: language)
+    input.map do |word|
+      stemmer.stem(word)
+    end
+  end
+  
+  # Counting the words, generate array for each element like => [word, frequency]
+  def self.count_words(input)
+    input.each_with_object(Hash.new(0)) { |token, hash| hash[token] += 1 }.sort_by {|k,v| v}.reverse
```

Comment:
I used the same line for word count before. And for sorting, since `value` is the last entry, you could use `sort_by(&:last).reverse` instead for simplicity.

Reply:
Wow! 💯  I did not know that we could do that! Updated. Coming...